osm2lanes/src/algorithm.rs

```diff
@@ -1,610 +1,238 @@
+    Rank::normal(TMode::All, LaneType::Shoulder),
+];
+
+fn travel_lane(
+    t: &TravelLane,
+    main: bool,
+    traffic_direction: Direction,
+) -> (LaneType, Direction, EnumSet<TurnDirection>) {
+    let turn_forward = t.forward.turn.get(TMode::All);
+    let turn_backward = t.backward.turn.get(TMode::All);
+    if let Some((turn_forward, turn_backward)) = turn_forward.zip(turn_backward) {
+        let forward_base = turn_forward.base();
+        if forward_base.is_some() && forward_base == turn_backward.base() {
+            return (LaneType::SharedLeftTurn, Direction::Fwd, EnumSet::new());
         }
     }
-}
 
-fn add_parking_lanes(fwd_side: &mut Vec<LaneSpec>, back_side: &mut Vec<LaneSpec>, tags: &Tags) {
-    let has_parking = vec!["parallel", "diagonal", "perpendicular"];
-    let parking_lane_both =
-        tags.is("parking:both", "lane") || tags.is_any("parking:lane:both", has_parking.clone());
-    let parking_lane_fwd = parking_lane_both
-        || tags.is("parking:right", "lane")
-        || tags.is_any("parking:lane:right", has_parking.clone());
-    let parking_lane_back = parking_lane_both
-        || tags.is("parking:left", "lane")
-        || tags.is_any("parking:lane:left", has_parking);
-
-    if parking_lane_fwd {
-        fwd_side.push(fwd(LaneType::Parking));
-    }
-    if parking_lane_back {
-        back_side.push(back(LaneType::Parking));
-    }
-}
+    for rank in RANKS {
+        let access_forward = rank.is_allowed(&t.forward.access);
+        let access_backward = rank.is_allowed(&t.backward.access);
 
-fn add_sidewalks_and_shoulders(
-    fwd_side: &mut Vec<LaneSpec>,
-    back_side: &mut Vec<LaneSpec>,
-    tags: &Tags,
-    cfg: &MapConfig,
-) {
-    // Workaround broken things like https://www.openstreetmap.org/way/523882355. Sidewalks on a
-    // footway don't make sense.
-    if back_side.is_empty()
-        && fwd_side.len() == 1
-        && matches!(fwd_side[0].lt, LaneType::SharedUse | LaneType::Footway)
-    {
-        return;
-    }
+        let dir = match (access_forward, access_backward) {
+            (true, false) => Direction::Fwd,
+            (false, true) => Direction::Back,
+            (true, true) => traffic_direction, // TODO: Both directions
+            (false, false) => continue,
+        };
 
-    if tags.is("sidewalk", "both") {
-        fwd_side.push(fwd(LaneType::Sidewalk));
-        back_side.push(back(LaneType::Sidewalk));
-    } else if tags.is("sidewalk", "separate") && cfg.inferred_sidewalks {
-        // TODO Need to snap separate sidewalks to ways. Until then, just do this.
-        fwd_side.push(fwd(LaneType::Sidewalk));
-        if !back_side.is_empty() {
-            back_side.push(back(LaneType::Sidewalk));
-        }
-    } else if tags.is("sidewalk", "right") {
-        if cfg.driving_side == DrivingSide::Right {
-            fwd_side.push(fwd(LaneType::Sidewalk));
+        let lane_type = if main || rank.lane_type != LaneType::Footway {
```

Comment:
Looks like `main` is used here. I don't understand the logic here, comment please?